security/pkg/server/ca/server.go

```diff
@@ -94,24 +100,58 @@ func (s *Server) CreateCertificate(ctx context.Context, request *pb.IstioCertifi
 		s.monitoring.GetCertSignError(signErr.(*caerror.Error).ErrorType()).Increment()
 		return nil, status.Errorf(signErr.(*caerror.Error).HTTPErrorCode(), "CSR signing error (%v)", signErr.(*caerror.Error))
 	}
+	certSigner = s.certSignerDomain + "/" + certSigner
 	respCertChain := []string{string(cert)}
 	if len(certChainBytes) != 0 {
 		respCertChain = append(respCertChain, string(certChainBytes))
 	}
+	var rootCert, rootCertFromMeshConfig []byte
 	if len(rootCertBytes) == 0 {
-		rootCert, err := util.FindRootCertFromCertificateChainBytes(cert)
+		rootCertFromCertChain, err := util.FindRootCertFromCertificateChainBytes(cert)
 		if err != nil {
 			serverCaLog.Errorf("failed to find root cert from signed cert-chain (%v)", err.Error())
 			s.monitoring.GetCertSignError(err.(*caerror.Error).ErrorType()).Increment()
 			return nil, status.Errorf(err.(*caerror.Error).HTTPErrorCode(), "failed to find root cert from signed cert-chain (%v)", err.(*caerror.Error))
 		}
+		caCertificates := s.ca.GetCACertificatesFromMeshConfig()
```

Comment:
nit: I wonder if there is a way to move all this logic into the RA code as part of the ca.sign() call? The existing logic in this function -> (a) extract the external signer that is supposed to sign the cert (b) find the correct CA cert that maps to the signer (c) verify that the cert is indeed signed by the CA Cert - this logic is highly specific to the RA code and makes the existing Istiod CA code confusing to read IMHO. Perhaps ca and ra servers could both implement a signWithCertChain that could return the entire cert-chain as a string?

Reply:
Yes, Thanks for the suggestion, Let me refactor this to scoped this to RA code base.

Reply:
moved to SignWithCertChain in RA code